ruby_programming/basic_ruby/nested_collections.md

```diff
@@ -0,0 +1,330 @@
+~~~
+
+If you try to access an index of a nested element that does not exist, it will raise an `NoMethodError`, because the nil class does not have a `[]` method. However, just like a regular array, if you try to access an index inside of the nested element that does not exist, it will return nil.
+
+~~~ruby
+teacher_mailboxes[3][0]
+#=> NoMethodError
+teacher_mailboxes[0][4]
+#=> nil
+~~~
+
+If you want a nil value returned when trying to access an index of a nested element that does not exist, you can use the `#dig` method. This method can also be used when accessing an index inside of the nested element that does not exist.
+
+~~~ruby
+teacher_mailboxes.dig(3, 0)
+#=> nil
+teacher_mailboxes.dig(0, 4)
+#=> nil
+~~~
+
+### Creating a new nested array
+Now that you have seen how to access values inside a nested array, we need to take a step back to look at creating a new nested array. In a previous lesson you were taught to create a new array, by calling the Array.new method with up to 2 optional arguments (initial size and default value), like `Array.new(3)` or `Array.new(3, 7)`. However, there is one major "gotcha" that is important to point out. According to the [documentation](https://ruby-doc.org/core-2.7.2/Array.html) the second optional argument, for the default value, should only be used with an immutable (unable to be changed) object such as a number, boolean value, or symbol. Using a string, array, hash, or other mutable object may result in confusing behavior because each default value in the array will actually be a *reference to* the same default value. Therefore, any change to **one** of the elements will change **all** of the elements in the array.
+
+To create an array of mutable objects, you will need to pass the default value to `Array.new` in a block, using curly braces, instead of the second optional argument. This creates *distinct copies* of the default value in the array, rather than references to the same default value.
+
+To see this for yourself, let's look at two examples. This first example uses the second optional argument for the default value.
+
+~~~ruby
+mutable = Array.new(3, Array.new(2))
+#=> [[nil, nil], [nil, nil], [nil, nil]]
+mutable[0][0] = 1000
+#=> 1000
+mutable
+#=> [[1000, nil], [1000, nil], [1000, nil]]
+~~~
+
+Changing the value of the first element in the first nested array, causes the first element to change in all three nested arrays! This same behavior will happen with strings, hashes, or any other mutable objects.
+
+Now, let's take a look at an example that omits the second optional argument and instead passes in the mutable value in a block.
+
+~~~ruby
+immutable = Array.new(3) { Array.new(2) }
+#=> [[nil, nil], [nil, nil], [nil, nil]]
+immutable[0][0] = 1000
+#=> 1000
+immutable
+#=> [[1000, nil], [nil, nil], [nil, nil]]
+~~~
+
+Changing the value of the first element in the first nested array does not cause the value to change in any other nested array.
+
+### Adding and Removing Elements
```

Comment:
Whats your thoughts on showing adding an array as a value?
`test_scores << [1,2,3,4]`

Just to show them how you would add to the 2d array structure already in place.

Reply:
That is a good idea and probably more useful than the example of just adding a number to the end like `test_scores.push(99)`. I will will update that!